Linux: Rewrite the event loop using calloop

witelokk opened the pull request:
This PR unifies the event loop code for Wayland and X11. On Wayland, blocking dispatch is now used. On X11, the invisible window is no longer needed.


Release Notes:

- N/A

kvark committed:
linux/x11: prioritize input in the event loop

kvark committed:
blade: defer frame acquisition to after the path rasterization

witelokk committed:
rewrite linux event loop using calloop

Co-authored-by: Tadeo Kondrak <me@tadeo.ca>

kvark commented on crates/gpui/src/platform/linux/x11/client.rs:
```diff
@@ -46,192 +59,217 @@ impl X11Client {
+
+        {
+            let xcb_connection = Arc::clone(&xcb_connection);
+            std::thread::spawn(move || loop {
```
With this change, we have an extra thread to poll events from the windowing system. It also means that each message has to go through one more step: first get from X11 into this thread, then get from this thread over the channel into the `calloop` event loop, and then be processed.

tadeokondrak replied:
See the comments in https://github.com/Smithay/smithay/blob/a04d9ac8441e19e87a6af6410d0f1ddbf4b293c5/src/utils/x11rb.rs for some reasoning for this choice

kvark reviewed (commented):
I'd like to have us on the same page about event loops. Technically speaking, we could have a platform-specific event loop (that's the current code in `main`), or we could have a platform-abstract event loop (like in `winit`). Having something in-between - an abstract-ish event loop that's only used for Linux - seems like a strange design choice.
Is the intent here to also use `calloop` on macOS and Windows eventually?

kvark reviewed (dismissed):
a few more concerns

mikayla-maki commented:
> Having something in-between - an abstract-ish event loop that's only used for Linux - seems like a strange design choice.

I think having a reasonably thin eventloop adapter that meshes with Zed's callback style and provides a bit of structure to other code we write could be good. And it would be nice to not have the invisible window. But that said, I haven't looked too deeply yet :)

kvark commented:
Discussed this [on discord](https://discordapp.com/channels/869392257814519848/1204679850208657418/1210840532381728808) and got a better understanding about how `calloop` fits in. I think it's a good direction, thanks @tadeokondrak for explanation.
Please rebase the PR and address the review points. I think X11 may start working then.

tadeokondrak committed:
linux/x11: Move from dedicated thread to fd source for event loop

tadeokondrak committed:
linux: Use None instead of Duration::MAX

tadeokondrak committed:
linux: Check for X11 events before event loop sleeps

tadeokondrak committed:
linux: Stop event loop directly

tadeokondrak committed:
linux: Remove now-empty LinuxPlatformState

tadeokondrak committed:
linux/x11: Draw only after other events processed

tadeokondrak committed:
linux: Call quit callback in the right place

witelokk committed:
Merge pull request #1 from tadeokondrak/calloop

linux/x11: Move from dedicated thread to fd source for event loop

witelokk committed:
make x11 work

Co-authored-by: Tadeo Kondrak <me@tadeo.ca>

witelokk committed:
wayland: fix fractional scaling

witelokk committed:
Merge remote-tracking branch 'upstream/main' into calloop

witelokk committed:
use calloop timer in key repeat

witelokk committed:
fix clippy

mikayla-maki committed:
Make background multi threaded

mikayla-maki committed:
WIP: Switch to using expose event for x11 rendering
TODO: Make sure we're only redrawing every so often, rather than every expose

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# Date:      Tue Feb 27 14:51:50 2024 -0800
#
# On branch calloop
# Changes to be committed:
#	modified:   crates/gpui/src/platform/linux/client.rs
#	modified:   crates/gpui/src/platform/linux/dispatcher.rs
#	modified:   crates/gpui/src/platform/linux/platform.rs
#	modified:   crates/gpui/src/platform/linux/wayland/client.rs
#	modified:   crates/gpui/src/platform/linux/x11/client.rs
#	modified:   crates/gpui/src/platform/linux/x11/window.rs
#
# Changes not staged for commit:
#	modified:   crates/gpui/src/platform/linux/dispatcher.rs
#	modified:   crates/gpui/src/platform/linux/x11/client.rs
#
# Untracked files:
#	.vscode/
#
# Next commands to do (2 remaining commands):
#    reword 52fe5825a xImplement a calloop based timer system
#    reword 7997396ac Adjust refresh rate based on monitor refresh
# You are currently editing a commit while rebasing branch 'calloop' on 'e2c9de0b0'.
#
# Changes to be committed:
#	modified:   crates/gpui/src/platform/linux/client.rs
#	modified:   crates/gpui/src/platform/linux/dispatcher.rs
#	modified:   crates/gpui/src/platform/linux/platform.rs
#	modified:   crates/gpui/src/platform/linux/wayland/client.rs
#	modified:   crates/gpui/src/platform/linux/x11/client.rs
#	modified:   crates/gpui/src/platform/linux/x11/window.rs
#
# Untracked files:
#	.vscode/
#

mikayla-maki committed:
Implement a calloop based timer system

mikayla-maki committed:
Adjust refresh rate based on monitor refresh
co-authored-by: julia <julia@zed.dev>

mikayla-maki committed:
Remove unused synchronization

co-authored-by: julia <julia@zed.dev>

mikayla-maki committed:
fmt

mikayla-maki committed:
Remove another mutex

mikayla-maki committed:
Fix clippy

kvark commented on crates/gpui/src/platform/linux/x11/window.rs:
```diff
@@ -268,31 +271,34 @@ impl X11WindowState {
     pub fn refresh(&self) {
-        let mut cb = self.callbacks.lock();
-        if let Some(ref mut fun) = cb.request_frame {
+        let mut cb = self.callbacks.borrow_mut();
+        if let Some(mut fun) = cb.request_frame.take() {
```
what was the motivation for doing this? did we end up calling a callback within a callback?
